timeseries/rfcs/0002-write-api.md

```diff
@@ -0,0 +1,475 @@
+#[derive(Debug, Clone)]
+pub struct TimeSeriesConfig {
+    /// Storage backend configuration
+    pub storage: StorageConfig,
+
+    /// How often to flush data to durable storage
+    pub flush_interval: Duration,
+
+    /// Maximum age of data to retain
+    pub retention: Option<Duration>,
+}
+
+impl Default for TimeSeriesConfig {
+    fn default() -> Self {
+        Self {
+            storage: StorageConfig::default(),
+            flush_interval: Duration::from_secs(60),
+            retention: None,
+        }
+    }
+}
+```
+
+#### Write API
+
+The write API consists of a single batched method with atomic semantics:
+
+```rust
+impl TimeSeries {
+    /// Write one or more time series.
+    ///
+    /// This is the primary write method. It accepts a batch of series,
+    /// each containing labels and one or more samples. The method returns
+    /// when the data has been accepted for ingestion (but not necessarily
+    /// flushed to durable storage).
+    ///
+    /// # Atomicity
+    ///
+    /// This operation is atomic: either all series in the batch are accepted,
+    /// or none are. This matches the behavior of `Log::append()`.
+    ///
+    /// # Series Identification
+    ///
+    /// Each unique combination of labels identifies a distinct time series.
+    /// The label set must include a `__name__` label for the metric name.
+    ///
+    /// # Ordering
+    ///
+    /// Samples within a series should be in timestamp order, but out-of-order
+    /// samples are accepted. Duplicate timestamps for the same series will
+    /// overwrite previous values.
+    ///
+    /// # Example
+    ///
+    /// ```rust
+    /// let series = vec![
+    ///     Series::builder("cpu_usage")
+    ///         .label("host", "server1")
+    ///         .sample(1700000000000, 0.75)
+    ///         .sample(1700000001000, 0.82)
+    ///         .build(),
+    ///     Series::builder("cpu_usage")
+    ///         .label("host", "server2")
+    ///         .sample(1700000000000, 0.45)
+    ///         .build(),
+    /// ];
+    ///
+    /// ts.write(series).await?;
+    /// ```
+    pub async fn write(&self, series: Vec<Series>) -> Result<()>;
+
+    /// Write with custom options.
+    ///
+    /// Allows control over durability guarantees and other write behaviors.
+    pub async fn write_with_options(
+        &self,
+        series: Vec<Series>,
+        options: WriteOptions,
+    ) -> Result<()>;
+
+    /// Force flush all pending data to durable storage.
+    ///
+    /// Normally data is flushed according to `flush_interval`, but this
+    /// method can be used to ensure durability immediately.
+    pub async fn flush(&self) -> Result<()>;
+}
+
+/// Options for the write operation.
+#[derive(Debug, Clone, Default)]
+pub struct WriteOptions {
+    /// Wait for data to be flushed to durable storage before returning.
+    /// Default: false (return immediately after buffering)
+    pub await_durable: bool,
```

Comment:
I'm curious if we need this for timeseries data - right now it's not super straightforward to get a handler to await durable with the batching mechanism I implemented. It would complicate the code with some annoying async book-keeping.

Reply:
That's fair - I think we can eventually support it. Maybe we'll wait to implement it until we have slatedb-wal factored out or a pluggable slate memtable so that we don't need to handle the synchronization between the buffer and slate's `await_durable` mechanism ourselves (it was a bit of a pain in slatedb to handle that and i really don't want to custom implement it if we can avoid it).